aten/src/ATen/native/cuda/FusedAdamKernel.cu

```diff
@@ -0,0 +1,228 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/Dispatch.h>
+#include <ATen/native/ForeachUtils.h>
+#include <ATen/native/cuda/ForeachFunctors.cuh>
+#include <ATen/native/cuda/MultiTensorApply.cuh>
+#include <c10/macros/Macros.h>
+#include <c10/util/irange.h>
+#include <vector>
+
+
+namespace at { namespace native {
+
+namespace {
+constexpr uint8_t kParamIdx = 0;
+constexpr uint8_t kGradIdx = 1;
+constexpr uint8_t kExpAvgIdx = 2;
+constexpr uint8_t kExpAvgSqIdx = 3;
+constexpr uint8_t kMaxExpAvgSqIdx = 4;
+
+template <typename scalar_type, typename opmath_t, int depth=4>
+C10_DEVICE __forceinline__ void adam_math(
+    scalar_type r_args[depth][kILP],
+    const float* step_count,
+    const double lr,
+    const double beta1,
+    const double beta2,
+    const double weight_decay,
+    const double eps,
+    const bool maximize,
+    const bool amsgrad,
+    const float* grad_scale_ptr,
+    const float* found_inf_ptr
+) {
+#pragma unroll
+    for (int ii = 0; ii < kILP; ii++) {
+        // Load values.
+        opmath_t param = static_cast<opmath_t>(r_args[kParamIdx][ii]);
+        opmath_t grad = static_cast<opmath_t>(r_args[kGradIdx][ii]);
+        if (grad_scale_ptr) {
+            grad /= (static_cast<double>(*grad_scale_ptr));
```

Comment:
do you need a double here? Also, is grad_scale_ptr exposed in public API? Please leave a comment that it's unused if it isn't. 
Edit: oh I see, some versions do support grad scaling

Reply:
I'm not sure but `GradScaler.unscale_` casts its scale to double when taking reciprocal:

https://github.com/pytorch/pytorch/blob/2f9d046d6717d9423e012c5249624fbeeb506cbb/torch/cuda/amp/grad_scaler.py#L277